File: android/src/main/kotlin/com/ileaf/flutter_stripe_identity/FlutterStripeIdentityPlugin.kt
package com.ileaf.flutter_stripe_identity

import android.content.ContentValues.TAG
import android.content.Context
import android.net.Uri
import android.util.Log
import androidx.activity.ComponentActivity
import com.stripe.android.identity.IdentityVerificationSheet
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlutterStripeIdentityPlugin */
class FlutterStripeIdentityPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var activity: ComponentActivity
    private lateinit var identityVerificationSheet: IdentityVerificationSheet
    private lateinit var identityVerificationResult: Result

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_stripe_identity")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> result.success("Android")
            "stripeIdentityVerification" -> {
                identityVerificationResult = result
                onIdentify(
                    call.argument<String>("verificationSessionId") ?: "",
                    call.argument<String>("ephemeralKeySecret") ?: ""
                )
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onDetachedFromActivity() {
        TODO("Not yet implemented")
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        TODO("Not yet implemented")
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity as ComponentActivity
        configureIdentitySheet(activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        TODO("Not yet implemented")
    }

    private fun configureIdentitySheet(activity: ComponentActivity) {
        identityVerificationSheet =
            IdentityVerificationSheet.create(
                activity, IdentityVerificationSheet.Configuration(
                    // Pass your square brand logo by creating it from local resource or
                    brandLogo = Uri.parse("https://loremflickr.com/cache/resized/65535_52750642974_d92f69439c_q_120_120_nofilter.jpg")
                )
            ) { verificationFlowResult ->
                when (verificationFlowResult) {
                    is IdentityVerificationSheet.VerificationFlowResult.Completed -> {
                        // The user has completed uploading their documents.
                        // Let them know that the verification is processing.
                        Log.d(TAG, "Verification Flow Completed!")
                        identityVerificationResult.success("Verification Flow Completed!")
                    }

                    is IdentityVerificationSheet.VerificationFlowResult.Canceled -> {
                        // The user did not complete uploading their documents.
                        // You should allow them to try again.
                        Log.d(TAG, "Verification Flow Canceled!")
                        identityVerificationResult.success("Verification Flow Canceled!")
                    }

                    is IdentityVerificationSheet.VerificationFlowResult.Failed -> {
                        // If the flow fails, you should display the localized error
                        // message to your user using throwable.getLocalizedMessage()
                        Log.d(TAG, "Verification Flow Failed!")
                        identityVerificationResult.error("0", "Verification Flow Failed!", null)
                    }
                }
            }
    }

    private fun onIdentify(verificationSessionId: String, ephemeralKeySecret: String) {
        // show loading UI
        // Request the session ID with Fuel or other network libraries
        // start verification session
        if (verificationSessionId.isNotEmpty() && ephemeralKeySecret.isNotEmpty()) {
            identityVerificationSheet.present(
                verificationSessionId = verificationSessionId,
                ephemeralKeySecret = ephemeralKeySecret
            )
        } else {
            identityVerificationResult.error("0", "Verification Flow Failed!", null)
        }
    }
}

File: ios/Classes/FlutterStripeIdentityPlugin.swift
import Flutter
import UIKit
import StripeIdentity

public class FlutterStripeIdentityPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_stripe_identity", binaryMessenger: registrar.messenger())
    let instance = FlutterStripeIdentityPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "stripeIdentityVerification":
        if let parameters: [String:Any] = call.arguments as? [String:Any] {
            
            var verificationSessionId = parameters["verificationSessionId"] as! String
            var ephemeralKeySecret = parameters["ephemeralKeySecret"] as! String
            presentVerificationSheet(verificationSessionId:verificationSessionId,ephemeralKeySecret:ephemeralKeySecret, callback: result)
        }
    default:
      result(FlutterMethodNotImplemented)
    }
    
  }
    func presentVerificationSheet(verificationSessionId: String, ephemeralKeySecret: String,callback:@escaping FlutterResult) {
  
    // Instantiate and present the sheet
      if #available(iOS 13, *) {
          // Configure a square brand logo. Recommended image size is 32 x 32 points.
          let configuration = IdentityVerificationSheet.Configuration(
              brandLogo: UIImage(systemName: "building.2.fill")!
          )
          let verificationSheet = IdentityVerificationSheet(
            verificationSessionId: verificationSessionId,
            ephemeralKeySecret: ephemeralKeySecret,
            configuration: configuration
          )
          verificationSheet.present(from:  UIApplication.topViewController() ?? UIViewController(), completion: { result in
            switch result {
            case .flowCompleted:
              // The user has completed uploading their documents.
              // Let them know that the verification is processing.
              print("Verification Flow Completed!")
                callback("Verification Flow Completed!")
            case .flowCanceled:
              // The user did not complete uploading their documents.
              // You should allow them to try again.
                callback("Verification Flow Canceled!")
              print("Verification Flow Canceled!")
            case .flowFailed(let error):
              // If the flow fails, you should display the localized error
              // message to your user using error.localizedDescription
              print("Verification Flow Failed!")
                callback("Verification Flow Failed!")
              print(error.localizedDescription)
            }
          })

      } else {
          // Fallback on earlier versions
      }
  }
}
extension UIApplication {
    class func topViewController(base: UIViewController? =  UIApplication.shared.windows.first { $0.isKeyWindow }?.rootViewController) -> UIViewController? {
        if let nav = base as? UINavigationController {
            return topViewController(base: nav.visibleViewController)
        }
        if let tab = base as? UITabBarController {
            if let selected = tab.selectedViewController {
                return topViewController(base: selected)
            }
        }
        if let presented = base?.presentedViewController {
            return topViewController(base: presented)
        }
        return base
    }
}
